### ISSUE TITLE
`<map>` suggested for `std::pair`

### ISSUE THREAD

**Issue by firewave** [include-what-you-use/include-what-you-use#1479]
```
/__w/cppcheck/cppcheck/lib/addoninfo.cpp should add these lines:
#include <map>      // for std::pair, std::operator!=, std::operator==
```

According to https://en.cppreference.com/w/cpp/utility/pair the symbol is provided by `utility`.

This started with (switching to) IWYU 0.21 from OpenSUSE. We previously used 0.20 from Fedora. So it might also be related to the distro.

As I have no local system with that version I do not have sample code at the moment.

**Comment by kimgr** [comment]
`std::pair` is attributed to `<utility>` with a reduced example similar to this: https://github.com/danmar/cppcheck/blob/bd1c5c66932278fd645d775a9a22c7c411d7ce45/lib/addoninfo.cpp#L60 on my Ubuntu+GCC11.4 system. I have some other noise around `std::_Rb_tree_iterator`, but `pair` actually looks good.

I think in general, the distro as a whole shouldn't matter much, but the installed GCC (or rather, libstdc++) version might.

In the example above it seems to me `std::map`, `std::pair` and others should be reexported by `picojson.h` via `json.h`, but that feels like a bigger challenge for IWYU to understand unprompted.

**Comment by firewave** [comment]
It looks like this is fixed in 0.22. I will try to create a reproducer with the older version so it can be properly verified what actually fixed this.

**Comment by firewave** [comment]
I now have binaries back to 0.20 locally and I cannot reproduce this.

But with the latest I get something strange:
```
/mnt/s/GitHub/cppcheck-fw/lib/addoninfo.cpp should add these lines:
[...]
#include <tr1/utility>  // for pair
```

The versions prior suggest `<utility>`.

**Comment by kimgr** [comment]
@firewave Good catch! Looks like there's a stray mapping from `bits/stl_pair.h` to `<tr1/utility>` now. I have a mapping refresh underway that should fix this.

**Comment by firewave** [comment]
> I have a mapping refresh underway that should fix this.

There are more mappings which need to be refreshed. I have some local changes I hope to publish soon.

**Comment by kimgr** [comment]
I'm looking at machine generating them, this file-by-file mapping is driving me nuts :-)

**Comment by kimgr** [comment]
This reduced repro:
```
// tests/1479/dumbjson.h
#include <map>
#include <utility>
#include <string>

typedef std::map<std::string, std::string> object_t;

object_t get_object() {
  return object_t();
}

// tests/1479/t.cc
#include "dumbjson.h"

void foo() {
  object_t o = get_object();
  auto it = o.find("blah");
  if (it != o.end()) {
    (void)it->second;
  }
}
```
now results in:
```
$ include-what-you-use tests/1479/t.cc 

tests/1479/t.cc should add these lines:
#include <map>         // for operator!=
#include <string>      // for basic_string, operator<
#include <utility>     // for pair

tests/1479/t.cc should remove these lines:

The full include-list for tests/1479/t.cc:
#include <map>         // for operator!=
#include <string>      // for basic_string, operator<
#include <utility>     // for pair
#include "dumbjson.h"  // for get_object, object_t
```
... which isn't perfect -- not sure why `<string>` is necessary for `operator<` -- but pair is definitely attributed to `<utility>` now.

Closing as fixed.

**Comment by firewave** [comment]
> ... which isn't perfect -- not sure why `<string>` is necessary for `operator<` -- but pair is definitely attributed to `<utility>` now.

I think that might be covered by #1651.
